### DISCUSSION TITLE
Yii 2.2

### DISCUSSION THREAD

**Discussion by bizley** [yiisoft/yii2#19831] [Yii 22]
Let's talk about possibility to bump Yii to ~2.1~ (2.2).

**Comment by bizley** [comment]
I would gladly help doing this although with my current work and duties I will not be able to pull it in reasonable time. Some help is needed definitely.

The things to consider are (among others of course):
- the minimum PHP version
- versioning (obviously v3 is taken so again no semver here)
- continuing the support of Yii 2.0

**Reply by tomaszkane** [reply]
IMHO as @schmunk42 says: https://packagist.org/packages/yiisoft/yii2/php-stats PHP **5.4** is 0.0 - 0.1% installations - so (almost) no one needs support this version.
For me resonable minimum is last "7" branch: **7.4**.

**Reply by lucaswitch** [reply]
@tomaszkane Yii capability of support so many php versions is actually awesome but i think it's time to move the minimum php version to 7.x, since that were so many good php changes.

**Reply by tomaszkane** [reply]
@lucaswitch Yii 2.2 already exist with PHP 8.1 as minimum version.
https://github.com/yiisoft/yii2/blob/3587dfb32249ea5a0fcf21d5bd0f98d296fd436a/framework/composer.json#L66

**Comment by lesha724** [comment]
https://github.com/yiisoft/yii2/discussions/19255

**Reply by lesha724** [reply]
https://github.com/yiisoft/yii2/discussions/19367

**Comment by diego-mathis** [comment]
I would like to see Yii2 being in sync with supported PHP versions, so that would be 8,1+
edit: actually php 8.0 since that is suported till end of year. https://www.php.net/supported-versions.php

**Reply by samdark** [reply]
What do you mean by "in sync"?

**Comment by xepozz** [comment]
I'd suggest to make a research and try to reuse Yii3 packages instead of Yii2 ones if it possible: some helpers, adapters for validators and etc.
I thought about that many times. Looks possible if we do it step by step.

**Reply by lesha724** [reply]
Yii3 packages has requirements php8.0 or higher

**Reply by tomaszkane** [reply]
Yii3 packages require policy is (AFAIK) "actually active/supported PHP version".

**Reply by xepozz** [reply]
It's not a problem to add BC layer for PHP 7.4 and fix dependencies.

**Reply by samdark** [reply]
That's a good thing to do and I've tried that already on several projects. Works wonders. But you have to go 8.1 first.

**Comment by rob006** [comment]
Please, do no use 2.1 as version, pick 2.2 instead - there was already a lot of fuss with 2.1, which had some planned features (and some of them implemented), then renamed to 3.0, and then abandoned. It will be so confusing is someone will try to search about 2.1 and find contradicting information. 

IMO it is very similar situation as with PHP 6.

**Reply by bizley** [reply]
Ok, this is the easiest of our concerns :) Yii 2.2 is ok.

**Reply by samdark** [reply]
Makes sense. Thanks.

**Reply by razvanphp** [reply]
Yii3 abandoned?!

**Reply by tomaszkane** [reply]
No. Why?

**Reply by razvanphp** [reply]
That’s how it sounded from the text above, but I guess what Rob meant was that 2.1 was abandoned.

**Reply by samdark** [reply]
Yes. 2.1 was abandoned back then. 2.2 is in preparation stage. Yii3 packages are mostly released along with docs. The guide and the rest of packages are in development. I'm running a few production instances with Yii3 components used.

**Comment by schmunk42** [comment]
Actually Yii2 is in code freeze state - but the reality is different :)

> the minimum PHP version

I'd advocate for `8.1` - I upgraded a lot of projects in the past months and while upgrading to 8.0 is rather easy, 8.1 and 8.2 raises a lot more of issues. So to avoid to be stuck inbetween versions, do the jump now to an active stable version.
Users should be advised to `require "2.0.x"` or add `conflict < Yii 2.2` for exisiting projects to avoid breaks.

> continuing the support of Yii 2.0

If there's a Yii 2.2 - for 2.0 CVEs and Bugs ONLY, see first line ;)

2.2 has to remain API compatible with 2.0.

**Reply by bizley** [reply]
> 2.2 has to remain API compatible with 2.0.

Only if we would be semver compatible and we are not.

**Reply by schmunk42** [reply]
What I wanted to say is, there should be an easy upgrade path to 2.2 for Yii2 users, but no active development in the classical sense, since I fear it would consume too much resources.

Active development should go in Yii3.

**Reply by bizley** [reply]
Agree. As for the code change I would for starters add types everywhere, and remove compatibility with older PHP versions. Then we can add features introduced by the new PHP versions as long as we can stay BC.

As for the min. PHP version - 8.1 is nice but do we support it "forever" again?

**Reply by terabytesoftw** [reply]
> Agree. As for the code change I would for starters add types everywhere, and remove compatibility with older PHP versions. Then we can add features introduced by the new PHP versions as long as we can stay BC.
> 
> As for the min. PHP version - 8.1 is nice but do we support it "forever" again?

I think not, the core must move forward with the language.

**Reply by rob006** [reply]
I suggest to take a look at https://packagist.org/packages/yiisoft/yii2/php-stats - 7.4 is still the most poplar PHP version for Yii 2 (and PHP 7 in general has around 50% of installs), so it does not make sense to jump straight to PHP 8+. We can always release 2.3 after few years with another required PHP version bump.

Also, I would be careful with adding types to signatures, as types in phpdoc are often incorrect. And without union types (PHP 8.0+ feature), we won't be able to represent types used by many methods anyway.

**Reply by bizley** [reply]
Yes, so it depends on the min. version indeed.

**Reply by schmunk42** [reply]
> 7.4 is still the most poplar PHP version for Yii 2 (and PHP 7 in general has around 50% of installs)

You can still use 2.0 with PHP 7.4.

> And without union types (PHP 8.0+ feature), we won't be able to represent types used by many methods anyway.

Isn't that a strong reason to move to PHP 8? :)

**Reply by rob006** [reply]
> Isn't that a strong reason to move to PHP 8? :)

IMO it is a strong reason not to add types blindly to method signatures :). 

In general, I think that even with union types it will be hard to add types in most of the places and we should limit such changes to places where there is no doubt what types are expected, for example accept only strings for string-related methods. Doing this for methods like `Query::where()` seems to be too risky IMO. 
Also, we can't really represent `string|string[]` using PHP types, and `string|array` is not that useful in that case (you can still pass `null` in array).

**Reply by schmunk42** [reply]
> IMO it is a strong reason not to add types blindly to method signatures :).

Sure ... but PHP has now been "infected" with strict-typing and once a lib or framework introduces it you can practically introduce it "everywhere" (except some edge cases as you mentioned). And Yii2 won't really work without union-types.

So regarding PHP 7.4, why support an already EOL version, when you can just use Yii 2.0.x for it?

**Reply by rob006** [reply]
From my experience it is way harder to introduce types in existing code that in a new one, and doing it in framework is even more tricky (Symfony started adding types in 2019, and it is still not finished). I'm not even sure if it is worth the trouble for project in feature freeze.

> So regarding PHP 7.4, why support an already EOL version, when you can just use Yii 2.0.x for it?

By cutting half of user base you increase chance, that some edge case will not be found before release. And adding additional type to method signature is a BC break, so you can't easily fix it after stable release.

**Reply by schmunk42** [reply]
I agree with your points. 
Adding types is an absolute pain in existing code, especially when you get newer (strict-typed) code in vendor.

Which real benefits would we get from moving from min requirement 5.4 to 7.4?
Could someone link some example issues?

**Reply by bizley** [reply]
> And adding additional type to method signature is a BC break, so you can't easily fix it after stable release.

So just code guards for phpdocs types where we can do it?

**Reply by tomaszkane** [reply]
> Which real benefits would we get from moving from min requirement 5.4 to 7.4? Could someone link some example issues?

I see only one: almost everyone who uses Yii2 can run it on his host with most popular PHP - 7.4. So we can use when it's possible 7.4 features and drop old/ugly code for < 7.4.

> You can still use 2.0 with PHP 7.4.
> Isn't that a strong reason to move to PHP 8? :)

I like this question :-) BTW, in my company several projects on Yii2 runs on PHP 8.0.

**Reply by schmunk42** [reply]
> [...] drop old/ugly code for < 7.4

Some changes which come to my mind
- `::class()` vs `::className` (was introduced in 5.5)
- Scalar type declarations (7.0) *not sure if we want this though*
- Null coalescing operator (7.0)

More stuff: https://web.archive.org/web/20201205203503/https://www.php.net/manual/en/appendices.php

But all of those are syntatic sugar in the core. Every change needs to be tested thoroughly, there were already esoteric issues with stuff like that.

> BTW, in my company several projects on Yii2 runs on PHP 8.0.

In mine too and upgrading to 8.0 was rather easy, but 8.1 and even more 8.2 were painful.

**Reply by mtangoo** [reply]
In our company we are running on 8.1 with no issues. 8.2 is rather a problem. We are waiting for a pending PR to be merged

**Reply by rhertogh** [reply]
I would also advocate for PHP 8.1. It's already supported in Yii 2.0, so anyone who need a lower PHP version can stay on that version and step by step transition to PHP 8.1. From there they could make the switch to Yii 2.2 if they want.

**Reply by nd4c** [reply]
Just a note.  I currently have an e-commerce/MIS site with an extensive backend (the MIS part) that runs on PHP 8.3.  I've extended some Yii classes to add in functionality and may have adjusted some things there to help with 8.3 - not sure right now - a lot of the code was written originally for Yii1 and has gone through a lot of upgrades.

I remember having some issues during php upgrades over the years - I'd vote for being more cutting edge and support at least 8.1.  PHP's version path moves rather quickly. (https://www.php.net/supported-versions.php)  Seriously, some of the posts here are over a year old - PHP has already EOL'd at least one if not two versions since this conversation started. ;-)

**Reply by schmunk42** [reply]
Now, I'd go with 8.3 - 8.2 active support ends in 1 month.

**Comment by samdark** [comment]
# Stats

<img width="117" alt="image" src="https://github.com/yiisoft/yii2/assets/47294/5915f1e5-3602-4179-b5ca-4be8263b45de">

- hhvm and PHP 5 are safe to drop support for.
- pre-7.4 versions are 19% together and that is signficiant. I wonder if these are using older Yii2 versions as well but we don't have the stats.
- 7.4 is the most popular version despite it being EOL-ed.

<img width="646" alt="image" src="https://github.com/yiisoft/yii2/assets/47294/8e01a39d-15b7-4ad9-8711-897d253979c6">

The tendency towards PHP 8 is there. It went to 50% in ~3 years.

# Resources

1. Focus of current team is Yii3 so we can't pull big upgrade w/o extra funding. You can see it with PHP 8.2 support. It goes too slow while, in reality, that's a one-man job for ~1 week + fixing edge cases (at least that was the case with Yii 1.1 that I did initially).
2. Community effort would definitely help but that still requires time to review/fix things. See point 1.

# Thoughts

- Dropping PHP 5 could be done in regular minor version. Composer will just install older version with matching constraint. That's tedious task but pretty much doable. That would help us a bit maintenence-wise.
- Adding types everywhere isn't going to play well with AR and is a huge effort. Way bigger than adding 8.2 support that is going really slow now.
- We can potentially drop PHP 7 w/o adding types and release Yii 2.2 along with providing a good upgrade guide and to ease maintenance significantly and overall pushing community to up to date versions.

**Reply by bizley** [reply]
I'm thinking right now about adding the guards in public/protected methods that will check the types of provided arguments and throw InvalidArgumentExc. This is is a bit of overhead but I think it's still ok-ish. This could be dropped when we reach min. PHP version with union types (still checking for array elements types though).

By looking at the stats I think it's safe to set min. PHP to 7.2 since older versions are 5.3% although based on general consensus it would be better to stick to 7.4.

**Reply by samdark** [reply]
@bizley that's even more work than adding types right away.

**Reply by bizley** [reply]
Hm, yes, maybe I would like to have more that we can do. It would make our lives simpler. But you are right.

**Comment by bizley** [comment]
As for the Yii 2.0 maintenance I would drop adding "enhancements" in total. Only security fixes and compatibility layer (which is problematic anyway).

**Reply by samdark** [reply]
Yes. That was an idea of "code freeze" years back...

**Reply by terabytesoftw** [reply]
This has turned into a snowball, the more versions of PHP are released, the more difficult it will be to have compatibility, so i think it should be done, with a minimum version of at least 8.0 or 8.1, whoever uses previous versions will stay at 2.0.

**Reply by bizley** [reply]
This is just changing the starting point of supporting the previous versions. PHP rather won't stop at 8.3.

**Comment by erickskrauch** [comment]
My vote is also for dropping old PHP versions.
But there is also a solution by using Rector with downgrade rules:
https://github.com/rectorphp/rector-downgrade-php

**Comment by samdark** [comment]
Some reasons on why people don't switch to PHP 8: https://twitter.com/sam_dark/status/1656549413282611200

**Reply by schmunk42** [reply]
> Some of the extensions in yii2 I rely on only work on 7.4 unfortunately :(

The biggest issue IMHO, when I see the current state how regularly extensions are maintained.

**Comment by terabytesoftw** [comment]
We could do the following.

- Create branch 2.2.
- Upload the minimum version of PHP to the framework and oficial extensions according to agreement.
- Pass the rector to all framework code and official extensions.
- Upload the version of PHPUNIT without any patch, and fix the tests, without even touching the code.

**Comment by rustamwin** [comment]
Is there any (major) plan to do in 2.2?

**Reply by samdark** [reply]
No.

**Comment by rhertogh** [comment]
In the discussion about which PHP version should be the minimum for Yii 2.2 the argument of current usage by PHP version is brought up a lot. But in my opinion there is a big flaw in that reasoning, it's missing the user needs. I think the current users can be grouped in 3 groups:

1. Users that want to move forward.
   - Willing to upgrade to Yii 2.2
   - Use a recent PHP version (>= 8.1)
2. Users that are happy to stay where they are.
   - Are OK with staying on Yii 2.0
   - Probably are using an older PHP version (e.g. 7.4)
3. Users that would want to use Yii 2.2 but can't.
   - Would like to upgrade to Yii 2.2
   - Stuck on an older PHP version (e.g. 7.4)

So my argument is that even a lot of the current Yii 2 installs are on PHP < 8.1 they are most likely in group 2 (happy to stay). Therefore the question should be: Should group 1 be hold back by group 3?
Since I don't think group 3 makes up a significant percentage I would argue that Yii 2.2 should focus on group 1 and use PHP 8.1 as the minimum version.

**Reply by smirnov-e** [reply]
Group 2, in my opinion, are so fine that they don't run `composer upgrade` and their code is not in active development anyway. So, as users of Yii 2.2, they can be considered inexistent.

My point is that most installs with old PHP > 7, < 8 should be from group 3 that want to upgrade, but can't do this. And this group should be rather large.

**Reply by nd4c** [reply]
On the symfony site, there is a page to help with installing the framework and has some technical requirements:
https://symfony.com/doc/current/setup.html#technical-requirements

And I quote from that page:
Technical Requirements
Before creating your first Symfony application you must: Install PHP 8.2 or higher...

We need to move on.  My guess (and experience) is that the hardest part to upgrading to a new PHP version is the underlying framework and libraries - not the project code.  That's "relatively easy" to correct.  I believe we need to start at the cutting edge and then keep moving forward.

**Reply by nd4c** [reply]
Just a comment on my previous post.  By "relatively easy" I mean that the programmers should be able to upgrade their own project.  Most IDE's can help with this.  And then just run your test suite - you'll find out where to fix it.

However, the programmers have very little control over the framework and the libraries.  Typically libararies have versions to help with supporting php versioning.  So Yii2.0 could support older php and Yii2.2 could support 8.2.  Yii2.2.2 (or something) could support 8.3, etc.  

Versioning is typically used for Security, Bug and Feature changes.  Support for various PHP versions should be considered as botha Security and Feature change.

**Comment by rustamwin** [comment]
It would be nice to cover the codebase with psalm.

**Comment by samdark** [comment]
Thought about it more. Current team doesn't have time to do it in a reasonable time but I see positive feedback about the idea itself. Therefore, I think we can form dedicated Yii2 team from community members who are willing to move it forward.

So if you want to be in this new team, let me know.

**Reply by simialbi** [reply]
@samdark I think I would have time to assist

**Reply by mtangoo** [reply]
We have a big Yii2 app and so if need be, I can help push it forward!

**Reply by rhertogh** [reply]
@samdark I'm also able and willing to help.

**Reply by samdark** [reply]
@simialbi, @mtangoo, @rhertogh gave you permissions to manage Yii2 issues and code. How would you want to communicate? We can do it here via discussions or via Telegram group or via something else.

**Reply by mtangoo** [reply]
Thanks @samdark 
I think Telegram is solid and available in all platforms. 
Of the public issues, I think here is a good place too!

I assume some veteran Yii dev(s) will be in the channel for guiding through!

**Reply by jiaweipan** [reply]
We have a big Yii2 app too! I willing to help. @samdark

**Reply by simialbi** [reply]
@samdark I think it's the best we communicate here, so we're sure no one has to install an additional app or something 😉. What's the plan now? Min PHP version 8.1? Or 7.4 to not exclude to much users (although it's already out of support https://www.php.net/supported-versions.php)

**Reply by rhertogh** [reply]
I would also prefer to keep the communication on Github. Then we have one central place and it's also easier to reference discussions and decisions.

**Reply by terabytesoftw** [reply]
I would like more 8.0 at least, to take advantage of language enhancements.

**Reply by samdark** [reply]
OK. Let's stick to GitHub. I've created a dedicated place: https://github.com/yiisoft/yii2/discussions/categories/yii-2-2

**Reply by samdark** [reply]
@jiaweipan since you're relatively new to dealing with Yii source code and pull requests, let's stick to usual PR workflow for now, OK?

**Reply by jiaweipan** [reply]
Fair enough.  I'm very happy to help it. @samdark

**Reply by themikem** [reply]
Available to help support this, and may have some resources that could help with specific tasks. Is there a punch list for 8.2 issues somewhere?

**Reply by mtangoo** [reply]
@themikem you can read 8.2 progress and needed help in 8.2 PRs but especially https://github.com/yiisoft/yii2/pull/19749

**Reply by themikem** [reply]
Ah okay, had the impression from discussions above that there was still outstanding work to do there to get even the basic code working, but it looks like its more just updating to 8.1 language features as a baseline at this point now that the referenced PR is merged?

(Though looking at 8.2 might also make sense given 8.1 is <6 months from end of active support...?)

**Reply by mtangoo** [reply]
8.2 still have some outstanding issues referenced in the merged PR and elsewhere. 
Surely any help to get them ironed is welcome!

**Reply by machour** [reply]
I'll be glad to resume my duties here, if there's a clear plan for what needs to be done. I'll be available at the end of July.

**Reply by samdark** [reply]
@machour added you back. No clear plan yet. @terabytesoftw is doing a rough draft and fixing tests to run smooth. After that is done, we can decide on the scope and distribute tasks.

**Comment by lourdas** [comment]
I read this thread and I'm a bit puzzled. What's the point of Yii 2.2? Just to raise the minimum PHP version and just make sure it works? Ok, so, what then?

In my organization, we started writing a webapp from scratch back in late 2011. We used Yii 1.1 at that time. The application is elections specific, so, it's not used all the time, but before and after the elections here in my country. The last time it was used was last June (12 years later) and it run without any issues in PHP 8.0. I tested it a bit with 8.1 and it seems to work fine too. Over the course of the years, the application has grown to several controllers and modules and it will take many months to convert it to Yii 2.0, or even more to another framework. So, unless something big happens, the application will stay in Yii 1.1 forever. Of course, since I know that it will run under 8.0 or 8.1, I can use many newer PHP features in my code and that's fine.

What's my point? If 2.2 is just about raising the minimum PHP version without major code refactorings in the framework itself, IMHO it's not a very good idea for a 2.2 branch. Just focus on the existing 2.0 branch, accept fixes for 8.1.x or 8.2.x and all will be good.

Please correct me if I misunderstood something.

**Reply by mtangoo** [reply]
Hi @lourdas 
I think PHP Version is prominent in this discussion because it is central to the very purpose of 2.2. This version is supposed to be a bridge to v3. It will be breaking BC with no apology and might even under the hood use Yii3 packages than Yii 2.0.x.

So it is going to be more than just PHP version change if it come to materialization. 

You are welcome to add your own opinions!

**Reply by terabytesoftw** [reply]
This change will also allow the following:

1.- Remove all depreciations.
2.- Clean all tests..
3.- And break BC without worrying about previous versions.
4.- This week i'm going to put some love into branch 2.2, and i'll advance it as much as possible.

**Reply by schmunk42** [reply]
> And break BC without worrying about previous versions.

Is this really the intended way to go concerning BC for 2.2?
IMHO this will limit the audience a lot and might be pretty hard to detect and test for existing applications ... and even more for 3rd party extensions.

My 2 cents.

**Reply by tomaszkane** [reply]
This is what it's all about, not to cling to old (unused-by-anyone) versions of PHP.

**Reply by schmunk42** [reply]
> versions of PHP.

I was referring to BC breaks in Yii2, not necessarily PHP.

**Reply by tomaszkane** [reply]
Yes, but raise minimum version of PHP is something like cause BC with some extension locked on lower than 8.2.

**Reply by rob006** [reply]
Increasing PHP version is not really a BC break, because it won't break existing code, new version just won't install on incompatible systems. Technically we could increase PHP requirements in 2.0.50 and people with old PHP will use 2.0.49.x just fine.

**Reply by lourdas** [reply]
Don't be so sure of that. When going from 7.4 to 8.0 for project of mine, I had to change a few lines because of deprecation notices. so it might not always be that easy.

**Reply by rob006** [reply]
> Don't be so sure of that. When going from 7.4 to 8.0 for project of mine, I had to change a few lines because of deprecation notices. so it might not always be that easy.

I mean increasing min PHP version required by framework - dropping some old PHP version won't add any additional depredations. And composer should prevent installing incompatible version of framework on environments using old PHP.

**Reply by lourdas** [reply]
Maybe I'm missing something that I don't get, but doesn't raising the minimum version mean that you must actually install a newer version of PHP so that the code runs? In this case, you should probably expect deprecations or even removals (if some eg. functions were already deprecated in the current PHP version that you didn't happen to be aware of), leading to non-working code.

**Reply by rob006** [reply]
You don't need to install latest version of framework. If you're using old PHP you can stick to Yii 2.0.49, since you don't need compatibility with PHP 8+. Yii 2.0.49 could be some kind of LTS with security fixes only.

**Reply by OndrejVasicek** [reply]
But staying in 2.0.49 would cut me from bugfixes in 51,52, etc, wouldn't it? Or you mean that that 2.0.49 would get fixes in 2.0.49.1, 2.0.49.2 etc? It sounds like very crazy versioning.

**Reply by tomaszkane** [reply]
It should be like now:
* 2.0.49
* 2.0.50
* 2.0.51...

and new branch/line:
* 2.2.0
* 2.2.1
* 2.2.2...

**Reply by rob006** [reply]
> But staying in 2.0.49 would cut me from bugfixes in 51,52, etc, wouldn't it?

Yes, But you would need to stay on 2.0.49 only if you want to stay on PHP 7.2 or lower.

**Comment by dneprovec** [comment]
So, this discussion came to nothing?

No chance to have a bit more modern Yii2 with which we are so familiar with?
A lot of good extensions and our personal code are wasted and cannot  be used in new projects.

Guys , you need to be "not lying" about such things.
Starting fake discussions , saw it in Samdark telegram, and then "you will have to do it yourself".

If your work with Yii3 came to fault, it would be reasonable to accept this reality and think about what Yii framework people could use in the current time.

**Reply by bizley** [reply]
WDYM?

**Reply by smirnov-e** [reply]
Well, it was talk about possibility. Not rock hard roadmap.
And this talk didn't came to nothing, as I see: https://github.com/yiisoft/yii2/pull/19955 (yes, it's just branch update, but it displays at least some interest from developers).

**Reply by terabytesoftw** [reply]
This branch will move forward, we will eliminate all the deprecations, and I will make some improvements, trying to break the code as little as possible.

**Reply by schmunk42** [reply]
> This branch will move forward, we will eliminate all the deprecations, and I will make some improvements, trying to break the code as little as possible.

Sounds good to me.

**Comment by terabytesoftw** [comment]
Hi, here is the repository for app-basic for Yii2, which contains:

1.- Vertical Slices design.
2.- Toggle for color mode (auto, dark, light), you can implement more colors in Bootstrap5.
3.- 100% code coverage.
4.- Static analysis with PHPStan.
5.- Check dependencies using composer-require-checker.
6.- Dependency injection usage.
7.- Utilizes Yii2 Branch 2.2, which requires at least PHP 8.1 and has no deprecated features.

You can find the repository here: https://github.com/yii2-extensions/app-basic

**Reply by terabytesoftw** [reply]
Hi, Web Application Advanced:

1️⃣ Vertical Slices design 📐
2️⃣ Color mode toggle (auto, dark, light) 🌗
3️⃣ 100% code coverage 📊
4️⃣ Static analysis with PHPStan Level 5 🕵️
5️⃣ Composer-require-checker for dependencies 🧩
6️⃣ Dependency injection usage 📥
7️⃣ Yii2 Branch 2.2, PHP 8.1+ 🐘
8️⃣ Example of Definitions & Singletons in Configuration 🛠️
9️⃣ yiisoft/config style Yii3 (entry scripts backend, console, frontend) 🧰

You can find the repository here: https://github.com/yii2-extensions/app-advanced

**Comment by razvanphp** [comment]
Can we still talk about features/roadmap for 2.2 here?

Besides the PHP upgrade, the most painful thing for me with yii2 is related to DevOps, deployment in particular: the AssetManager needs to be refactored in a way to make stateless deployments possible, in docker containers, specifically.

Not sure what is the best way, but one idea I tried is [mikk150/yii2-asset-manager-flysystem](https://packagist.org/packages/mikk150/yii2-asset-manager-flysystem), or maybe we should include a command that publishes all assets from CLI.

I'm willing to work on this after we brainstorm ideas. Thank you!

**Reply by mtangoo** [reply]
@razvanphp Yes sure. one of those was something to do with assets management. @terabytesoftw was leading on that.
Can you open new discussion on how you want Yii2.2 to work with assets so that we can have full discussion there?

**Reply by razvanphp** [reply]
a new discussion thread? Do we expect such a big debate? not sure if I have much more to add than the two suggestions above.

**Reply by mtangoo** [reply]
I do not know how long or short it will be. But imagine every single feature discussed in this one thread. It will be messy.
As of the suggestions, they need to be looked at from different angles, pros and cons discussed, PR made may be et al.
All that is not as easy as you seem to suggest.

**Reply by mtangoo** [reply]
@razvanphp feel free to share your ideas at https://github.com/yiisoft/yii2/discussions/20164

**Comment by nd4c** [comment]
This conversation seems to have dwindled.  I realize the world keeps us all very busy, and as far as Yii2 goes, I find that it still works very well.  It's just a little old and some of the coding practices have changed.

I'd like to share some thoughts.  Maybe it will help.  Please forgive me if I offend anyone (I tend to do that...)

There are a lot of frameworks out there, some with very impressive backing.  But Yii2 has always been near the top (and some would say at the top) for years.  It's an impressive framework.  But why?  

Other than the great coding and thought that went into it, it offers some compelling advantages:

[ ] It is easy to roll out a new project with just about everything you needed to support a robust website: MCV, Security, Auth, Database, Ajax, an easy templating systems (Views), etc.
[ ] It has an impressive testing framework.
[ ] It is very easy to adjust and re-organize the code to how you want it.
and the list goes on.

Yii3 on the other hand is cutting edge, uses the latest coding patterns and is extremely moldable.  I just find it hard to start up a new project.  There are way too many decisions to make and some of the info on how to add functionality is difficult to find and/or understand. In short there is an incredibly large learning curve.  This is very different from Yii2.

Laravel is more popular than Symfony for one simple reason: It's easy to get started and has a smaller learning curve.  Both are impressive frameworks and are very similar to Yii3 in using the latest modular coding patterns - they are SOLID.  With Laravel you can start a new project and add in basic security (auth) with just a couple of commands.  Symfony, on the other hand, has a BOOK to describe how to do it.  Phew!  (And in the book the author makes all kinds of decisions for you - kind of defeating the foundation of the framework.)

I propose that Yii2.2 be simply a prefactored, ready-to-go, implementation for Yii3.  And, like Yii2, with it's basic and advanced versions, offer a selection of "Project Types".

To be continued...?

**Comment by nd4c** [comment]
a continuation...

I'm sometimes called a full-stack programmer.  I sometimes called a lot of bad things as well, but, anyway...

I remember when Centos first came out and it was a Godsend!  So stable!  Before that it was a constant battle to update the servers.  You always had to have a test server to try out anything before updating a production server.  Just the server, never mind the code.

And the same with PHP.  Trying to mix and match various libraries was a nightmore!  Even with PEAR.  Then along came Composer!  Another Godsend!

And many years ago I'd written several website with just PHP - no libraries.  In fact, the one time I used a library - the site got hacked because of a hole in the library and brought down the server - crash!  Then along came PHP Frameworks!  Yes, another Godsend!

My point?  What we need as programmers is solid supporting infrastructure.  The less we have to worry about it the better.

That's why I went with Yii 1.  It gave a solid foundation for my websites, and I could spend most of my time writing the business logic and having fun with front end layout.  Very nice!  Yii2 was even better.  (Well, the migration from Yii1 to Yii2 was intense - but worth it.)

So I'd like to recommend that we de-emphasize the modularity of Yii3 and wrap it up a bit with "ready to run" implementations.  It will always be modular.  But as programmers I think we'd like something a little more ready to go for most websites.   

Symfony is fantastic, but having to go through a whole Book just to get a robust working website is too much.  I'd like to hope that Yii3 would not require such a book...

**Comment by Renkas** [comment]
Has the 2.2 development stalled completely? There's only three open issues. Is there any chance that it will be released at all?

**Reply by mtangoo** [reply]
No. It is a huge undertaking. In due time it will get updates. Patience is virtue here 😀

**Reply by samdark** [reply]
@terabytesoftw is preparing some code to be reviewed/merged. His branch diverged too much from what we have in the branch, unfortunately.

**Reply by xepozz** [reply]
I've also made a bridge to the https://github.com/yiisoft/yii-debug and https://github.com/yiisoft/yii-dev-panel
Early alpha, but still

https://github.com/yiisoft/yii2-debug/pull/533

**Comment by terabytesoftw** [comment]
I've been working on the `2.2` branch locally, although there's no development on the branch yet, so far we have the following ROADMAP that we'll be doing in the next few days:

- [x] Update to the latest master branch. 
- [ ] Apply `Rector` to clean up dead code from versions previous to `PHP 8.1`.
- [ ] Enable `StyleCI` for the `2.2` branch if possible.
- [ ] Integrate `PHPStan` for static analysis, configuring a baseline to allow progressive improvements over time.
- [ ] Once the branch is clean, we will make an initial release of the `2.2` branch along with all official extensions, providing a stable starting point without breaking changes.
- [ ] Gradually apply improvements and review pending PRs to determine what can be safely merged into `2.2`.

**Reply by schmunk42** [reply]
How about asset management aka remove fxp?

**Reply by terabytesoftw** [reply]
If this is the first change i'll make, and i'll add support for Vite, or i could use any asset manager, npm, bun, or foxy extension, etc.

**Reply by lisotton** [reply]
I see that the last commit in the branch `2.2` is from 6 months ago. What is the status of the 2.2 release?

**Reply by terabytesoftw** [reply]
The updated branch is `22.0`, you can try it, the same branch number exists in all extension repositories.

**Comment by terabytesoftw** [comment]
This discussion will be closed. 

If you have any suggestions, you can open a new one in the Yii 2.2 category. https://github.com/yiisoft/yii2/discussions/categories/yii-2-2
